### PR TITLE
create opcode uses deterministic evm contract address

### LINKED ISSUE #426: bug: the EVM address of a deployed contract is wrong
# Bug Report

**Current behavior:**

The EVM address of a contract account is computed as follows:

```
        // Generate EVM_contract address from the new cairo contract
        // TODO: Use RLP to compute proper EVM address, see https://www.evm.codes/#f0
        let (_, low) = split_felt(starknet_contract_address);
        local evm_contract_address = 0xAbdE100700000000000000000000000000000000 + low;
```

**Expected behavior:**

Depending on wether the call is a `CREATE` or a `CREATE2`, the address should be computed with

- CREATE (see https://www.evm.codes/#f0)
```
address = keccak256(rlp([sender_address,sender_nonce]))[12:]
```
- CREATE2 (see https://www.evm.codes/#f5)
```
address = keccak256(0xff + sender_address + salt + keccak256(initialisation_code))[12:]
```

**Related code:**

See `ContractAccount.deploy` function.
RLP utils are added in https://github.com/sayajin-labs/kakarot/pull/422

### PR DESCRIPTION
**Pull request by jobez**
Time spent on this PR: 1/2 day

## Pull request type

Please check the type of change your PR introduces:

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Create opcode uses an easter egg 🥚 to determine the evm contract mapping.

Resolves #426 

## What is the new behavior?

The create opcode uses `CreateHelper.get_create_address` which derives the address via:

```
address = keccak256(rlp([sender_address,sender_nonce]))[12:]
```

-
-
-

### CODE REVIEW

File: src/utils/rlp.cairo [hunk 1]
```diff
@@ -140,4 +141,86 @@ namespace RLP {
...
+            // single byte needs no prefix
+            assert [rlp + rlp_len] = element;
+            return (rlp_len + 1,);
+        } else {
+            // otherwise, we need to convert the value to an unpadded byte array
+            let (local reversed_element_bytes: felt*) = alloc();
+            let (local element_bytes: felt*) = alloc();
+
+            let (element_high, element_low) = split_felt(element);
+            let (element_bytes_len) = Helpers.uint256_to_bytes_no_padding(
```
Comment (jobez) [inline src/utils/rlp.cairo:170]: when testing against the case where the salt was sufficiently large [255,` 2**55-1), the Helpers.felt_to_bytes_array was returning incorrect results.

This motivated creating this helper that encoded a uint256 with no padding for RLP encoding.

File: src/utils/utils.cairo [hunk 2]
```diff
@@ -236,6 +243,29 @@ namespace Helpers {
         return (bytes_array_len=32, bytes_array=value_as_bytes_array);
     }
 
+    // @notice This function is a variant of `uint256_to_bytes_array` that encodes the uint256 with no padding
+    // @param value: value to convert.
+    // @param idx: index of res array
+    // @param res: resultant encoded bytearray, but in reverse order
+    // @param dest: reversed res, putting byte array in right order
+    // @return bytes array len
+    func uint256_to_bytes_no_padding{bitwise_ptr: BitwiseBuiltin*, range_check_ptr}(
```
Comment (jobez) [inline src/utils/utils.cairo:252]: To reiterate the motivation for this helper:

We encode the salt/nonce in rlp, which means translating it to a byte array when its value is greater than 127.

When I tested the salt in the case 2^55-1, `Helpers.felt_to_bytes_array` was returning a byte array of length one with [2^55-1], downstream causing incorrect rlp encoding than the python implementation.

So I opted to lift the value as a uint256 to handle a greater range accurately and needed to write this helper to get the byte array that took the uint256 value and translated it into a byte array with no padding.

File: tests/unit/src/kakarot/instructions/test_system_operations.py [hunk 3]
```diff
@@ -71,8 +71,9 @@ async def test_call(
             contract_account_class.class_hash, account_registry.contract_address
         ).call()
 
+    @pytest.mark.parametrize("salt", [127, 256, 2**55 - 1])
```
Comment (jobez) [inline tests/unit/src/kakarot/instructions/test_system_operations.py:74]: 127 would fit in a single byte, meaning in rlp terms we wouldn't need a prefix.

256 and 2**55-1 both need prefixes

File: src/utils/rlp.cairo [hunk 4]
```diff
@@ -140,4 +141,86 @@ namespace RLP {
             return (rlp_len=data_len + 1);
         }
     }
+
+    // @notice encodes felt into an rlp element
+    // @param element The felt that is encoded into rlp
+    // @param rlp_len The length of the rlp array
+    // @param rlp The pointer receiving the rlp encoded felt
+    // @return rlp_len The length of the encoded felt in bytes
+    func encode_element_felt{
```
Comment (jobez) [inline src/utils/rlp.cairo:150]: motivation: to encode the salt/nonce value in rlp element (which has distinct logic from an rlp list).

see: https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp/

**Review by jobez — COMMENTED** [review]
Currently in draft to have some overview of the motivation/logic of the PR.

File: pyproject.toml [hunk 5]
```diff
@@ -29,6 +29,7 @@ eth-account = "<0.8.0"
 eth-keys = "<0.4.0"
 autoflake = "^2.0.0"
 pycryptodome = "^3.16.0"
+rlp = "<3"
```
Comment (jobez) [inline pyproject.toml:32]: seems as if i need to commit an updated lock file

the output of the error is saying to run

```
poetry lock --no-update
```

Is it as simple as that and then pushing the updated lock file?

File: src/utils/rlp.cairo (hunk 4 above)
Comment (ClementWalter) [reply src/utils/rlp.cairo:150]: could be named just `encode_felt`?

File: src/utils/rlp.cairo [hunk 6]
```diff
@@ -140,4 +141,86 @@ namespace RLP {
...
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        bitwise_ptr: BitwiseBuiltin*,
+        range_check_ptr,
+    }(element: felt, rlp_len: felt, rlp: felt*) -> (rlp_len: felt) {
+        alloc_locals;
+        // 127 is the largest value that can be represented by one byte
+        let element_is_single_byte = is_le(element, 127);
```
Comment (ClementWalter) [inline src/utils/rlp.cairo:158]: is `element` an "item" (as per your link)? if yes, then plz name it `item`

File: src/utils/rlp.cairo (hunk 1 above)
Comment (ClementWalter) [reply src/utils/rlp.cairo:170]: we have a terrible naming imho from the very beginning regarding `bytes`, `padding` etc.

We should strive to use the same naming convention as the EVM, ie:
- word = `bytes32` = `uint256`
- `bytes_i_` is the _i_ first bytes from the left:
    - `bytes31(bytes32)` will drop the right most byte
    - `bytes1(bytes32)` will keep the left most byte
- `uint_i_` works the opposite (as uint are big endian)
    - for example, with no overflow, `uint64(uint256)` = `uint64`
    - so `uint248(uint256)` will drop the left most byte
    - `uint8(uint256)` will keep the right most byte

`bytes` is a `bytes1[]`, ie an array of bytes and consequently should not have any leading or trailing 0s added for padding

In short, I think that we should once for all create a `BytesLib`, put all these helpers with standard naming conventions and tests, and stop to think about it!

File: src/utils/rlp.cairo [hunk 7]
```diff
@@ -140,4 +141,86 @@ namespace RLP {
...
+    // @return rlp_len The length of the encoded list in bytes
+    func encode_element_byte_array{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        bitwise_ptr: BitwiseBuiltin*,
+        range_check_ptr,
+    }(byte_array_len, byte_array: felt*, rlp_len: felt, rlp: felt*) -> (rlp_len: felt) {
+        alloc_locals;
+
+        let element_is_le_55 = is_le(byte_array_len, 55);
```
Comment (ClementWalter) [inline src/utils/rlp.cairo:197]: `item` ?

File: tests/unit/src/kakarot/instructions/test_system_operations.py [hunk 8]
```diff
@@ -71,8 +71,9 @@ async def test_call(
...
     async def test_create(
-        self, system_operations, contract_account_class, account_registry
+        self, system_operations, contract_account_class, account_registry, salt
```
Comment (ClementWalter) [inline tests/unit/src/kakarot/instructions/test_system_operations.py:76]: you don't use `salt` in the test 🤔

**Review by ClementWalter — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 5 above)
Comment (ClementWalter) [reply pyproject.toml:32]: no, it's easier 😇

actually when you want to add a lib, you don't (never) update the `pyproject.toml` by hand. Instead, you want `poetry` (or the other package manager you use) to make it for you as it'll check dependencies versions consistency.

So you should use instead

```
poetry add "rlp<3" --group dev
```

and it'll be update the `pyproject.toml`, the `poetry.lock` file and install the lib in your local interpreter.

Comment (jobez) [reply pyproject.toml:32]: thank you!

File: src/utils/rlp.cairo (hunk 1 above)
Comment (jobez) [reply src/utils/rlp.cairo:170]: Are you ok with `Helpers.uint256_to_bytes_no_padding` as a name for now?

File: src/utils/utils.cairo [hunk 9]
```diff
@@ -236,6 +243,29 @@ namespace Helpers {
         return (bytes_array_len=32, bytes_array=value_as_bytes_array);
     }
 
+    // @notice This function is a variant of `uint256_to_bytes_array` that encodes the uint256 with no padding
+    // @param value: value to convert.
```
Comment (jobez) [inline src/utils/utils.cairo:247]: convention question:

What is the preferred format for these document comments? In some places it looks like:

https://github.com/sayajin-labs/kakarot/blob/main/src/kakarot/stack.cairo#L55

and here I went with the flow and used colons.

File: src/utils/rlp.cairo (hunk 1 above)
Comment (ClementWalter) [reply src/utils/rlp.cairo]: can you add an issue for this refactoring?

File: src/kakarot/instructions/system_operations.cairo [hunk 10]
```diff
@@ -551,9 +552,58 @@ namespace CreateHelper {
         pedersen_ptr: HashBuiltin*,
         range_check_ptr,
         bitwise_ptr: BitwiseBuiltin*,
-    }(sender_address: felt, nonce: felt) -> (evm_contract_address: felt) {
-        // TODO: see https://github.com/sayajin-labs/kakarot/issues/426
-        return (0xAbdE100700000000000000000000000000000000 + nonce,);
+    }(sender_address: felt, salt: felt) -> (evm_contract_address: felt) {
```
Comment (jobez) [inline src/kakarot/instructions/system_operations.cairo:555]: I thought I would preserve our back and forth here about potential next steps re: the salt usage here.

> me: just to sanity check my understanding, in the case of create, the evm documentation refer to a nonce. in our codebase we have salt storage variable. 

> these terms are pretty much interchangeable?

> clement: actually I think that the nonce is sender related while our salt is currently global
> actually I think that the nonce is sender related while our salt is currently global

File: src/utils/utils.cairo (hunk 9 above)
Comment (ClementWalter) [reply src/utils/utils.cairo:247]: we are supposed to use natspec https://docs.soliditylang.org/en/v0.8.17/natspec-format.html

could you create an issue for cleaning the codebase and adding a tool like https://github.com/onlydustxyz/kaaper to generate the doc from the comment?

File: src/kakarot/instructions/system_operations.cairo (hunk 10 above)
Comment (ClementWalter) [reply src/kakarot/instructions/system_operations.cairo:555]: yes, let's keep like that (`salt`)  and create an issue for investigating the usage of a true sender's nonce

File: src/utils/rlp.cairo (hunk 1 above)
Comment (jobez) [reply src/utils/rlp.cairo]: np! see #457

**Review by ClementWalter — APPROVED** [review]
